ThrusterOS, Diagnostics, GNC

DSsoto opened the pull request:
This is a revamp of the thruster related software on the sub.
Some of the commits from this week have not been tested on the sub due to the fact that it is disassembled. Let's wait for these things to get tested before we actually merge this.

Since this PR has a lot of changes it would be good to start getting it reviewed.

There are breaking changes to sim related thruster code. I am working on those now.

I would lay out the main changes here, but the commit messages are more complete.

DSsoto committed:
THRUSTERS: add calibration for VideoRay props

* renames calibration file to json_nez_prop.json
* updates launch file to use new calibration

DSsoto committed:
THRUSTERS: standardize Bmatrix thruster directions

* Make all of the thruster directions point from the motor housing to
  the shroud (this is the direction of force when given positive effort)
  (Documented in the sub map in the wiki)

DSsoto committed:
THRUSTERS: change sim thrust bounds

* Changes reflect switch to old VideoRay props

DSsoto committed:
THRUSTERS: version control firmware configuration

* Firmware configuration file names correspond to motor_id in the thruster driver.
* 'maximum_allowable_power' was set to 600 Watts for all thrusters
* Motor polarity was set up to coincide with vecotrs in the thruster_layout in
  sub8_thruster_mapper and the sub map on the wiki.

DSsoto committed:
GNC: make changes for debugging gnc issues

* thruster_driver.py:
  - Disable acting on bus voltage kills because of buggy voltage estimate
    introduced in #247
  - Set margin factor to 1.0 so that thusters can be utilized to their limit
    if necessary
* mapper.py:
  - Fix frame for /wrench_actual msg so it could be visualized in rviz
* Refactor config data into calibration and firmware directories

DSsoto committed:
THRUSTERCOMM: add all VideoRay CSR definitions

DSsoto committed:
THRUSTERS: add undeclared, unavailable exceptions

DSsoto committed:
THRUSTERSPINNER: add individual thrust control

* Allow for individual control as well as batch control

DSsoto committed:
THRUSTERS: fix race conditions for comms on port

* Implement a mutex for multiple thrusters' comms on a port
* Monitor latency of serial comms on per thruster basis

DSsoto committed:
THRUSTERPORT: load calibration by thruster port

* Each ThrusterPort will hold the effort to thrust mapping for all of
  the thrusters on the port

DSsoto committed:
THRUSTERMAPPER: get thrust bounds independently

* Add position and direction to ThrusterInfo service
* Make mapper get data for each thruster independently

DSsoto committed:
THRUSTERS: switch to 4th order polynomial calib

* Thruster calibration is now parametrized as a two 4th order polynomial
  mapping from thrust to effort (one for going forward and one for
  going backward)
* Make thrust bounds more conservative for all thrusters

DSsoto committed:
THRUSTERS: add individual feedback, refactor driver

* Publish thruster feedback on individual topics
* Extract bus voltage monitoring to a separate class
* Implement separate software deactivation of thrusters
* Only check for thrusters becoming unresponsive when
  a thrust command is issued
* Warn user when thrusters outside of bounds are commanded
  by the mapper
* Refactor driver to make it simpler and less asynchronous

DSsoto committed:
THRUSTERPORT: refactor ThrusterPort object

* Monitor thruster comms latency
* Monitor comms loss at the ThrusterPort level
* Keep track of which thrusters were declared but are not responsive
* Hold thruster model (pose, calibration, bounds) at the ThrusterPort
  level

DSsoto committed:
THRUSTERS: implement reading a single VR register

DSsoto committed:
THRUSTERS: implement setting a single VR register

DSsoto committed:
THRUSTERS: add rebooting VR thruster from software

* Thrusters don't need to be physically unplugged to reboot

DSsoto committed:
THRUSTERPORT: extract VideoRay CSR comms functions

* Extracted standalone VRCSR functions:
  + send_VRCSR_response_packet
  + parse_VRCSR_request_packet

DSsoto committed:
THRUSTERS: read 1 register w/o grabbing all memory

* Cleaned up inefficient register reading method written because of
  undocumented VideoRay comms functionality
* Refactored read and write register methods

DSsoto committed:
THRUSTERS: check for motor_id w/o sending thrust

* Refactor ThrusterPort.command_thruster

DSsoto committed:
THRUSTERS: add virtual Unfail service

DSsoto committed:
THRUSTERS: update calibration based on profiling

* All lateral thrusters were put on a load cell on 06-17. The outpputs
  were deemd consistent enought to fit a 2nd order polynoial to all
  together (0th order term zeroed out)

DSsoto committed:
THRUSTERS: add power to ThrusterStatus msg

* Also update msg field names to match CSR field names

DSsoto committed:
THRUSTERS: implement CONFIDENT_MODE

* Makes it so that command packets can continue to be sent to thrusters even if
  thrusters are found to be unresponsive, thruster-loss is never raised

RustyBamboo commented on gnc/sub8_thruster_mapper/nodes/mapper.py:
```diff
@@ -76,16 +78,18 @@ def get_ranges(self):
+        maxima = np.array([thruster_info_service_proxy(x).max_force for x in self.thruster_name_map])
+
+        #minima = np.array([thruster_range.min_force] * self.num_thrusters)
+        #maxima = np.array([thruster_range.max_force] * self.num_thrusters)
```
Dead code?

DSsoto replied:
Removed

RustyBamboo commented on gnc/sub8_thruster_mapper/nodes/mapper.py:
```diff
@@ -131,6 +135,7 @@ def map(self, wrench):
             - Account for variable thrusters
         '''
         thrust_cost = np.diag([1E-4] * self.num_thrusters)
+        print 'bounds: {}'.format(zip(self.min_thrusts, self.max_thrusts))
```
print statement

DSsoto replied:
Removed

RustyBamboo commented on gnc/sub8_thruster_mapper/config/thruster_layout.yaml:
```diff
@@ -1,66 +1,105 @@
   }
   BLV: {
-    motor_id:  5,
+    motor_id:  8,
```
Why 8?

DSsoto replied:
Was to lazy to reconfigure thruster, fixed.

RustyBamboo commented on gnc/sub8_thruster_mapper/config/thruster_layout.yaml:
```diff
@@ -1,66 +1,105 @@
 # motor_id's should be globally unique among all thrusters
-# all vectors should be specified in body frame
+# position and orientation vectors should be specified in body frame
+# calib contains the 4th to 1st order coefficients for a polynomial
```
I think it would be good to write the polynomial... Unless it's already written somewhere and I missed it

DSsoto replied:
Added it in a comment.

DSsoto committed:
THRUSTERS: add comms monitoring to thruster status

* The published thruster status msg after each command response
  is received now includes command_tx_count, status_rx_count and
  command_latency_avg

DSsoto committed:
GNC: update controller gains

* More conservative gains being used for debugging thruster issues

DSsoto committed:
THRUSTERS: improve error reporting, bugfix

* Always report node_name in thruster-out alarms
* Report errors regarding thruser fault conditions
* Don't check for thrusters if serial line not available
* Fix bug in thruster driver when updating the failed thruster list

DSsoto committed:
GNC: restore gains from THAT WEDNESDAY

* These are the "Feeling good on a wednesday gains" from my
  tuning session with Jasoun on June 7th.

DSsoto committed:
GNC: update thruster model to a linear one

* All the higher order terms were zeroed out for all thrusters.
* The linear term was chosen as the least squares fit to a second
  order model from our latest calibration that closely matched the data.

DSsoto committed:
ALARMHANDLERS: update kill and thruster-out alarms

* Handler now provides the initial kill alarm for the alarm server (killed)
* Handler now provides initial thruster-out alarm to alarm server that
  includes the 'offline_thruster_names' parameter

DSsoto committed:
ODOM: implement grace period for odom kill raise

DSsoto committed:
GNC: calibrate thrusters at lower power max

* Calibrated all of the thruster models using a power limit of 200 W
  (as opposed to 600 W before)
* Also updated the thrust bounds for the thrusters at this power limit

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/config/firmware_settings/3.txt:
```diff
@@ -0,0 +1,40 @@
+14: Positive duty limit:         0.990000
+15: Acceleration rate:           0.000050
+16: Deceleration rate:           0.000050
+17: Maximum allowable power:     600.000000
```
So we have them running at 200 max watts now? We should update these files then.

DSsoto replied:
Done

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/launch/thruster_driver.launch:
```diff
@@ -1,4 +1,4 @@
 <launch>
-    <node pkg="sub8_videoray_m5_thruster" type="thruster_driver.py" name="thruster_driver"
-        args="--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration.json" />
+  <param name='thrusters/calibration_dir' type='string' value='$find subb8_videoray_m5_thruster)/config/calibration/' />
```
don't think we use this param any more

DSsoto replied:
Restructured, removed individual launches in sub8_videoray and sub8_thruster_mapper and only thrusters.launch in sub8_launch.

RustyBamboo commented on drivers/sub8_videoray_m5_thruster/launch/thruster_driver.launch:
```diff
@@ -1,4 +1,4 @@
 <launch>
-    <node pkg="sub8_videoray_m5_thruster" type="thruster_driver.py" name="thruster_driver"
-        args="--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration.json" />
+  <param name='thrusters/calibration_dir' type='string' value='$find subb8_videoray_m5_thruster)/config/calibration/' />
```
extra b in subb8_videoray_m5...

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/readme.md:
```diff
@@ -11,10 +11,11 @@ It is now maintained by David Soto.
+The thrusters take an input on [-1, 1]. The files provided  in `config/calibration/` map commanded effort values to expected output thrust in Newtons.
 
-Run with the argument `--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration.json`, or the path to the appropriate calibration file
+Run with the argument `--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration/video_ray_prop_calibration.json`, or the path to the appropriate calibration file.
```
This is no longer accurate right? It reads the 4-th order coefficients from the thruster_layout file right?

DSsoto replied:
Updated

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/nodes/thruster_driver.py:
```diff
@@ -33,279 +37,238 @@ def __init__(self, window_duration):
 class ThrusterDriver(object):
     _dropped_timeout = 1.0  # s
     _window_duration = 30.0  # s
+    _NODE_NAME = 'thruster_driver'
```
Probably better to get this from [rospy](http://wiki.ros.org/rospy/Overview/Names%20and%20Node%20Information)

DSsoto replied:
Done

kev-the-dev commented on command/sub8_alarm/alarm_handlers/kill.py:
```diff
@@ -10,6 +10,9 @@ class Kill(HandlerBase):
 
     def __init__(self):
+        # Alarm server wil set this as the intial state of kill alarm (starts killed)
+        self.initial_alarm = Alarm(self.alarm_name, True, node_name='alarm_server')
```
Should set problem description to 'initially raised' or something so we know why it's raised.

DSsoto replied:
Done

kev-the-dev commented on command/sub8_launch/launch/subsystems/thrusters.launch:
```diff
@@ -1,5 +1,6 @@
-        <rosparam file="$(find sub8_thruster_mapper)/config/thruster_layout.yaml" param="thruster_layout"/>
-        <include file="$(find sub8_thruster_mapper)/launch/thruster_mapper.launch"/>	
-        <include file="$(find sub8_videoray_m5_thruster)/launch/thruster_driver.launch"/>
+  <rosparam file="$(find sub8_thruster_mapper)/config/thruster_layout.yaml" param="thruster_layout"/>
```
This is already set in the thruster mapper launch.

DSsoto replied:
Fixed

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/readme.md:
```diff
@@ -11,10 +11,11 @@ It is now maintained by David Soto.
-Run with the argument `--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration.json`, or the path to the appropriate calibration file
+Run with the argument `--calibration_path=$(find sub8_videoray_m5_thruster)/config/calibration/video_ray_prop_calibration.json`, or the path to the appropriate calibration file.
 
+Note: right now only one calibration can be used for all of the thrusters. A current goal is to implement per thruster calibration.
```
This is now implemented correct? If so update the readme

DSsoto replied:
Updated

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/sub8_thruster_comm/thruster_comm.py:
```diff
@@ -1,84 +1,181 @@
+            return serial_port
+        except serial.serialutil.SerialException as e:
+            rospy.logwarn("Could not connect to thruster port {}".format(port_name))
+            raise SubjuGatorException()
```
This would be a good example to add a SubjuGatorException inherited Exception class, like SubjuGatorSerialException

DSsoto replied:
I agree, done.

kev-the-dev commented on utils/sub8_diagnostics/sub8_exception/subjugator_exception.py:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python
+
+
+class SubjuGatorException(BaseException):
```
This was added in a separate PR right? I think you got to git it out

DSsoto replied:
Done

kev-the-dev reviewed (changes requested):
Good work. What I would really like to see is added test cases to the thruster unit test to test thing like failed thrusters (that alarm raises, thrust doesn't get published)

DSsoto committed:
THRUSTERS: add effort and thrust to status msg

kev-the-dev reviewed (commented):
Patch to update bagging variables to get the new thruster status messages
```
diff --git a/scripts/bagging_variables.sh b/scripts/bagging_variables.sh
index 5208c16..eb924b2 100755
--- a/scripts/bagging_variables.sh
+++ b/scripts/bagging_variables.sh
@@ -28,6 +28,6 @@ export bag_blueview_mono="/blueview_driver/image_mono"
 export bag_blueview="$bag_blueview_ranges $bag_blueview_color $bag_blueview_mono"
 export bag_controller="/pd_out /rise_6dof/parameter_descriptions /rise_6dof/parameter_updates"
 export bag_navigation="/wrench /wrench_actual /c3_trajectory_generator/trajectory_v /c3_trajectory_generator/waypoint /trajectory"
-export bag_thrusters="/thrusters/thrust /thrusters/thruster_status"
+export bag_thrusters="/thrusters/thrust /thrusters/status/FLV /thrusters/status/FLH /thrusters/thrust/FRV /thrusters/thrust/FRH /thrusters/thrust/BLV /thrusters/thrust/BLH /thrusters/thrust/BRV /thrusters/thrust/BRH"
 export bag_gnc="$bag_controller $bag_navigation $bag_thrusters"
 export bag_kill="$bag_gnc $bag_front_left_cam $bag_blueview"
```

DSsoto committed:
GNC: publish mapper wrench error

kev-the-dev reviewed (changes requested):
At testing 7/8 we found the thruster driver will not notice if thrusters come back online if an entire node was out. We need to refactor so this is noticed. You could:
* publish thrust=0 for offline thrusters in thruster mapper
* handle this as a special case in thruster driver

kev-the-dev reviewed (changes requested):
I also noticed in testing this branch that the FakeThrusterPort no longer works (several attributes are missing). We should still be able to run with /simulate True and at least see what the mapper produces

DSsoto committed:
THRUSTERS: use crc32 check for response validity

DSsoto committed:
SUB8EXCEPTION: implement repr and str methods

DSsoto committed:
MAPPER: send 0 thrust from dropped thrusters

DSsoto committed:
THRUSTERPORT: add save settings function

DSsoto committed:
BAGGING: add bag_thrusters to bagging_aliases

Credit: patch provided by ironmig

DSsoto committed:
THRUSTERS: add a thruster debug shell

* Also adds the thruster_debug_shell alias for it and the
  thruster_spinner alias for the sub8_diagnostics spinner tool.

DSsoto committed:
DIAGNOSTICS: implement thruster monitoring tmux

* TMUX session made to monitor thruster statuses.
* Session made to start automatically with the thruster launch
* Alias 'thruster_monitor' made for users to attach to session

DSsoto committed:
THRUSTERS: add function to enter config mode

DSsoto committed:
THRUSTERS: reverse engineer trigger csr addresses

* Fixed incorrectly reported csr addresses obtained from VideoRay
  documentation.

DSsoto committed:
THRUSTERS: add set registers from dict function

DSsoto committed:
THRUSTERS: transform motor settings file to yaml

DSsoto committed:
THRUSTERS: validate thrust command response crc32

DSsoto committed:
GNC: pool 7-13 edits for stationkeeping

DSsoto committed:
ALARMS: add bus-voltage handler that warns all

DSsoto committed:
MISC: address PR #290 comments

kev-the-dev reviewed (commented):
Also, please apply this diff to fix thruster_visualizer after your changes to the thruster layout
```
diff --git a/utils/sub8_diagnostics/scripts/thruster_visualizer.py b/utils/sub8_diagnostics/scripts/thruster_visualizer.py
index b127984..2165c91 100755
--- a/utils/sub8_diagnostics/scripts/thruster_visualizer.py
+++ b/utils/sub8_diagnostics/scripts/thruster_visualizer.py
@@ -23,7 +23,7 @@ class ThrusterVisualizer(object):
         assert self.layout is not None, 'Could not load thruster layout from ROS param'
 
         '''
-        Create MarkerArray with an arrow marker for each thruster at index motor_id.
+        Create MarkerArray with an arrow marker for each thruster at index node_id.
         The position of the marker is as specified in the layout, but the length of the arrow
         will be set at each thrust callback.
         '''
@@ -42,7 +42,7 @@ class ThrusterVisualizer(object):
             self.markers.markers.append(m)
         for key, layout in self.layout.iteritems():
             # Set position and id of marker from thruster layout
-            idx = layout['motor_id']
+            idx = layout['node_id']
             pt = numpy_to_point(layout['position'])
             self.markers.markers[idx].points.append(pt)
             self.markers.markers[idx].points.append(pt)
@@ -64,7 +64,7 @@ class ThrusterVisualizer(object):
             if cmd.name not in self.layout:  # Don't draw marker if thruster is not in layout
                 continue
             layout = self.layout[cmd.name]
-            idx = layout['motor_id']
+            idx = layout['node_id']
             bounds = layout['thrust_bounds']
 
             # Select an arrow length based on commanded thrust, max thrust (from layout), and the MAX_LENGTH constant
```

kev-the-dev commented on gnc/sub8_thruster_mapper/nodes/mapper.py:
```diff
@@ -76,16 +79,15 @@ def get_ranges(self):
-        rospy.logwarn("Waiting for service {}".format(range_service))
-        rospy.wait_for_service(range_service)
-        rospy.logwarn("Got {}".format(range_service))
+        thruster_info_service = 'thrusters/thruster_info'
```
This works, but seems unnecessary when all of this information is in the thruster ros param it already has and saves the service calls.

DSsoto replied:
Done

kev-the-dev commented on scripts/bash_aliases.sh:
```diff
@@ -29,3 +29,8 @@ alias subfmt="python2.7 -m flake8 --ignore E731 --max-line-length=120 --exclude=
 alias submodel="rosparam set /robot_description -t \$CATKIN_DIR/src/SubjuGator/command/sub8_missions/tools/sub8/sub8.xml"
+
+# Thrusters
+alias thruster_debug_shell="rosrun sub8_diagnostics thruster_debug_shell.py -i"
```
The convention for the rest of our aliases are to start them with sub. We should probably follow that for these as other MIL projects have thrusters

DSsoto replied:
Used prefix

kev-the-dev commented on utils/sub8_diagnostics/scripts/thruster_debug_shell.py:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env ipython
+# flake8: noqa
```
Why are we not format checking these files? I think there should always be a comment explaining why a file/line is ignored from flake8

DSsoto replied:
Added comment

kev-the-dev commented on command/sub8_alarm/alarm_handlers/bus_voltage.py:
```diff
@@ -0,0 +1,23 @@
+
+    def raised(self, alarm):
+        if alarm.severity == 3:
+            subprocess.Popen('echo "Warning: bus voltage is low!" | wall', shell=True,
```
From a glance at the code, it looks like this will spam the terminal as the alarm is raised continually. Either on the thruster driver end or the alarm handler end it should probably only wall (or raise) once for each of the warning and killing voltage

DSsoto replied:
Fixed

kev-the-dev commented on gnc/sub8_thruster_mapper/config/thruster_layout.yaml:
```diff
@@ -1,66 +1,105 @@
+# node_id's should be globally unique among all thrusters
+# position and orientation vectors should be specified in body frame
+# calib contains the 4th to 1st order coefficients for a polynomial
+#     mapping from thrust in Newtons to effort (e = t[0]x^4 + t[1]x^3 + t[2]x^2 + t[3]x)
```
Add a comment here explaining our temporary solution to have inaccurate high bounds / low calib numbers just to get things working. It would be confusing to read this in the future without this comment

DSsoto replied:
No longer necessary.

kev-the-dev reviewed (commented):
This is getting close to mergeable. Got a few more picked nits.

The failing test also has to get resolved. If you want to just disable it, I think that would be acceptable given our time constraints

DSsoto committed:
THRUSTERS: set all firmware settings from file

* Settings will be set a startup from files in the config directory

DSsoto committed:
VISUALIZATION: fix thruster visualizer after thruster layout changes

DSsoto committed:
GNC: pool 7-15 gnc tuning and bugfixes

* Result was great station keeping with small amount of wonkyness for
  long strafing moves
